subiquitycore/geoip.py

```diff
@@ -0,0 +1,116 @@
+# Copyright 2018-2021 Canonical, Ltd.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import asyncio
+import logging
+import requests
+from xml.etree import ElementTree
+
+from subiquitycore.async_helpers import (
+    CheckedSingleInstanceTask,
+    run_in_thread,
+    TaskFailure,
+)
+
+log = logging.getLogger('subiquitycore.geoip')
+
+
+class GeoIP:
+    """query geoip for CountryCode, TimeZone, other useful things"""
+
+    # sample:
+    #   <Response>
+    #     <Ip>1.2.3.4</Ip>
+    #     <Status>OK</Status>
+    #     <CountryCode>US</CountryCode>
+    #     <CountryCode3>USA</CountryCode3>
+    #     <CountryName>United States</CountryName>
+    #     <RegionCode>CA</RegionCode>
+    #     <RegionName>California</RegionName>
+    #     <City>Rio Vista</City>
+    #     <ZipPostalCode>94571</ZipPostalCode>
+    #     <Latitude>38.1637</Latitude>
+    #     <Longitude>-121.7016</Longitude>
+    #     <AreaCode>707</AreaCode>
+    #     <TimeZone>America/Los_Angeles</TimeZone>
+    #   </Response>
+
+    def __init__(self, app):
+        self.app = app
+        self.element = None
+        self.response_text = ''
+        self.valid = False
+        self.lookup_task = task = CheckedSingleInstanceTask(self._lookup)
+        self.app.hub.subscribe('network-up', task.start)
+        self.app.hub.subscribe('network-proxy-set', task.start)
+
+    async def lookup(self):
+        await self.lookup_task.start()
+        return self.valid
+
+    async def _lookup(self):
+        if self.valid:
+            return
+        await self.request_geoip_data()
+        self.load_element()
+        self.valid = True
+        self.app.hub.broadcast('geoip-data-ready')
```

Comment:
so in line with my above rambling, i'd rather this returned a value rather than stashing it on self.

Reply:
#983